webrender/src/render_task.rs

```diff
@@ -303,53 +341,54 @@ impl RenderTask {
     pub fn new_mask(
         key: Option<ClipId>,
         task_rect: DeviceIntRect,
-        raw_clips: &[ClipWorkItem],
-        extra_clip: Option<ClipWorkItem>,
+        raw_clips: ClipChain,
+        extra_clip: ClipChain,
         prim_rect: DeviceIntRect,
         clip_store: &ClipStore,
         is_axis_aligned: bool,
+        prim_coordinate_system_id: CoordinateSystemId,
     ) -> Option<RenderTask> {
         // Filter out all the clip instances that don't contribute to the result
+        let mut current_coordinate_system_id = prim_coordinate_system_id;
         let mut inner_rect = Some(task_rect);
-        let clips: Vec<_> = raw_clips
-            .iter()
-            .chain(extra_clip.iter())
-            .filter(|work_item| {
+        let clips: Vec<_> = ClipChainNodeIter { current: raw_clips }
+            .chain(ClipChainNodeIter { current: extra_clip })
+            .filter_map(|node| {
+                let work_item = node.work_item.clone();
+
+                // We cannot apply any kind of optimization based on inner rectangles when the
+                // clips exists in another axis-aligned coordinate system.
+                if node.work_item.has_compatible_coordinate_system(current_coordinate_system_id) {
```

Comment:
I don't think this is correct. Inner rects are valid regardless of the coordinate systems.

Reply:
This is just a replacement for the workaround in https://github.com/servo/webrender/pull/1820. Would it be better for me to add a FIXME and open an issue?

Reply:
yes, please